server/src/main/java/org/opensearch/common/cache/tier/TieredSpilloverCache.java

```diff
@@ -45,6 +47,7 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V>, StoreAwareCache
     private final Optional<StoreAwareCache<K, V>> onDiskCache;
```

Comment:
Can we move `TieredSpilloverCache` into the plugin?

The plugin can just produce an implementation of `ICache`.

Reply:
I don't think we should move this ```TieredSpilloverCache``` into plugin.
You can plugin any implementation of onHeap or onDisk into this tiered cache which makes it suitable to be kept within core itself as it is generic/flexible.

In case there is a case where for example, ehcache provides a tiered implementation with a combination of (onHeap + onDisk), then it makes sense for that to be loaded from plugin itself, by doing something like

```
Map.of(CacheStoreType.TIERED, new EhcacheTieredCache.Factory()) // something like that.
```

Reply:
It's a question of what we're abstracting, versus what we're setting in stone.

I was imagining that we would abstract "a cache". You could plug in whatever cache implementation you want (tiered, sharded, on-disk, in-memory (on- or off-heap), transported via carrier pigeon, etc).

In this implementation, we're saying that tiered caching is a core feature. Tiered caching prescribes an on-heap cache and an on-disk cache. The implementations of **those** caches can be provided by plugins, but we're absolutely committed to a) tiering with spillover, b) the tiers are on-heap and on-disk.

Reply:
>In this implementation, we're saying that tiered caching is a core feature.

I think we also need to reason that what should go in the plugin and what should not. I feel tiered spillover is generic enough(without any dependency) to be used within core considering we can use it within IndicesRequestCache/QueryCache.

Ehcache disk cache made sense to be moved within plugin considering it brought an external dependency.

I agree that we should have a way to plugin different kind of caches. Which is possible via the interface```StoreAwareCache``` we have already defined inside core. 

In case we want to use some other flavour of tiered cache, we can plug that in, otherwise we can default to TieredSpilloverCache(with plugged in required dependencies). Similar to Cache.java we have within OpenSearch which is the internal cache.

We can do something like

```
"indices.request.cache.onheap.store.name": "default" (Uses Cache.java implementation by default)
"indices.request.cache.onheap.store.name": "caffeine" (Uses caffeine instead loaded from plugin)

"indices.request.cache.tiered.store.name": "default" (Uses TieredSpilloverCache or anything inside core)
"indices.request.cache.tiered.store.name": "ehcache" (Uses ehcache version of tiered cache)
```

Reply:
To make sure we're on the same page, here are the two approaches from what I can see:

![Tiering_in_core](https://github.com/opensearch-project/OpenSearch/assets/346194/d52c3aea-088a-4780-b6c3-3d488401bf2d)

![Tiering_in_plugin](https://github.com/opensearch-project/OpenSearch/assets/346194/d16738a2-21f4-4d48-9baf-eb489afdf6b2)

I see no use for the `StoreAwareCache` interface. The TieredSpilloverCache just needs to know which cache to hit first, and which to hit second (or third, or fourth, etc.. as I alluded to [here](https://github.com/opensearch-project/OpenSearch/pull/11874#discussion_r1467009046)).

Is that a fair representation?

Reply:
A fair representation! 

My approach was around the first diagram where TieredSpilloverCache/OpenSearchOnHeapCache sits inside Core.

```TieredSpilloverCache``` currently implements high level interface ie ```ICache```.  And I had created ```StoreAwareCache``` for specific store types onDisk/onHeap.

I think I can change ```TieredSpilloverCache``` to implement StoreAwareCache instead as it is too store aware, just it can have multiple ones.
```
TieredSpilloverCache implements StoreAwareCache {
   CacheStoreType getStoreType() {
       return CacheStoreType.TIERED;
   }
}
```
CacheStoreType will have another type
```
CacheStoreType {
   DISK,
   ON_HEAP,
   TIERED
}
```

This way it kinda makes this design complete. Now anyone(like IndicesRequestCache) within OpenSearch can configure desired kind of cache store type via setting. And we can use above classes/enums to achieve that.

```
indices.requests.cache.type: on_heap (or disk or tiered) // Value is of type CacheStoreType.
```

Reply:
Right -- only I'm saying that I don't like the top diagram. I like the bottom diagram. 😁

Reply:
Regarding 
>I see no use for the StoreAwareCache interface. 

This is needed to determine relevant stats or source from where the data was cached or was evicted from. We get that from this interface.
For example, within TieredSpilloverCache, we use it to determine from which source the removal event happened and what kind of actions we need to perform. 
For example, a eviction happened from ON_HEAP storeType, and we spillover to DISK storeType.

Reply:
> This is needed to determine relevant stats or source from where the data was cached or was evicted from. We get that from this interface.
> For example, within TieredSpilloverCache, we use it to determine from which source the removal event happened and what kind of actions we need to perform.

Right, so it's an implementation detail for TieredSpilloverCache, which doesn't belong in core. I still don't think TieredSpilloverCache needs to work that way anyway, though, since TieredSpilloverCache can just define tier 1, tier 2, tier 3, etc. For stats, we can track them by cache name (as an implementation detail of TieredSpilloverCache). 

> Mainly the difference is with TieredSpilloverCache sitting inside a plugin.

Look at the first message from this thread. Michael Froh said "Can we move `TieredSpilloverCache` into the plugin?" It's literally the topic of discussion.

> It is similar to OpenSearchOnHeapCache in a way which also sits inside core.

The difference is that OpenSearchOnHeapCache is **required** to maintain existing behavior. We could also move OpenSearchOnHeapCache into a module, so that all cache implementations are plugged in. I would be happy with that approach too.

TieredSpilloverCache is optional, opt-in behavior. It's a different cache implementation. The separation of abstractions is "Core needs to know about caches, and interacts with caches through an interface. The cache implementations are provided through plugins/modules."

Reply:
I think you were using plugin/module interchangeably so I got confused.  😁
I agree that we can consider to move this into a "module" with OS, I was just not sure about the "plugin" part.

I can work on that to do this change.